subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildInclusionCoordinator.java

```diff
@@ -62,70 +93,52 @@ public void prepareRootBuildForInclusion() {
         }
     }
 
     public void registerSubstitutionsAvailableFor(BuildState build) {
         if (build instanceof RootBuildState) {
             registerGlobalLibrarySubstitutions();
         } else {
-            makeSubstitutionsAvailableFor(build, new HashSet<>());
+            withLockForBuild(build, () -> makeSubstitutionsAvailableFor(build));
         }
     }
 
     public void registerSubstitutionsProvidedBy(BuildState build) {
         if (build instanceof RootBuildState && registerRootSubstitutions) {
             // Make root build substitutions available
-            doRegisterSubstitutions((RootBuildState) build);
+            withLockForBuild(build, () -> substitutionRegistry.registerSubstitutionsFor((RootBuildState) build));
         }
     }
 
     public void prepareForPluginResolution(IncludedBuildState build) {
-        synchronized (registering) {
-            if (!registering.add(build)) {
-                return;
-            }
-            try {
-                build.ensureProjectsConfigured();
-                makeSubstitutionsAvailableFor(build, new HashSet<>());
-            } finally {
-                registering.remove(build);
-            }
-        }
+        withLockForBuild(build, build::ensureProjectsConfigured);
+        makeSubstitutionsAvailableFor(build);
     }
 
-    private void makeSubstitutionsAvailableFor(BuildState build, Set<BuildState> seen) {
-        // A build can see all the builds that it includes
-        if (!seen.add(build)) {
-            return;
-        }
-        synchronized (registering) {
-            boolean added = registering.add(build);
-            try {
-                for (IncludedBuildInternal reference : build.getMutableModel().includedBuilds()) {
-                    BuildState target = reference.getTarget();
-                    if (!registering.contains(target) && target instanceof IncludedBuildState) {
-                        doRegisterSubstitutions((IncludedBuildState) target);
-                        makeSubstitutionsAvailableFor(target, seen);
-                    }
-                }
-            } finally {
-                if (added) {
-                    registering.remove(build);
+    private void makeSubstitutionsAvailableFor(BuildState build) {
+        Deque<BuildState> stack = new ArrayDeque<>();
+        Set<BuildState> seen = new HashSet<>();
+
+        stack.push(build);
+        seen.add(build);
+
+        while (!stack.isEmpty()) {
+            BuildState current = stack.pop();
+
+            for (IncludedBuildInternal reference : current.getMutableModel().includedBuilds()) {
+                BuildState child = reference.getTarget();
+
+                if (seen.add(child) && child instanceof IncludedBuildState) {
+                    withLockForBuild(child, () -> substitutionRegistry.registerSubstitutionsFor((IncludedBuildState) child));
```

Comment:
💅  Given that the method body is already non-trivial, I would re-add the `doRegisterSubstitutions` method that was before and use it here and in other places. It would improve the readability.